# SVD for Movie Recommendations
In this notebook, I'll detail a basic version of model-based collaborative filtering for recommendations by employing it on the MovieLens 1M dataset.

[In my previous attempt](https://github.com/khanhnamle1994/movielens/blob/master/Content_Based_and_Collaborative_Filtering_Models.ipynb), I used user-based and item-based collaborative filtering to make movie recommendations from users' ratings data. I can only try them on a very small data sample (20,000 ratings), and ended up getting pretty high Root Mean Squared Error (bad recommendations). Memory-based collaborative filtering approaches that compute distance relationships between items or users have these two major issues:

1. It doesn't scale particularly well to massive datasets, especially for real-time recommendations based on user behavior similarities - which takes a lot of computations.
2. Ratings matrices may be overfitting to noisy representations of user tastes and preferences. When we use distance based "neighborhood" approaches on raw data, we match to sparse low-level details that we assume represent the user's preference vector instead of the vector itself.

Thus I'd need to apply **Dimensionality Reduction** technique to derive the tastes and preferences from the raw data, otherwise known as doing low-rank matrix factorization. Why reduce dimensions?

* I can discover hidden correlations / features in the raw data.
* I can remove redundant and noisy features that are not useful.
* I can interpret and visualize the data easier.
* I can also access easier data storage and processing.

With that goal in mind, I'll introduce Singular Vector Decomposition (SVD) to you, a powerful dimensionality reduction technique that is used heavily in modern model-based CF recommender system.

![dim-red](images/dimensionality-reduction.jpg)

## Loading the Dataset
Let's load the 3 data files just like last time.

In [27]:
# Import libraries
import numpy as np
import pandas as pd

# Reading ratings file
ratings = pd.read_csv('ratings.csv', sep='\t', encoding='latin-1', usecols=['user_id', 'movie_id', 'rating', 'timestamp'])

# Reading users file
users = pd.read_csv('users.csv', sep='\t', encoding='latin-1', usecols=['user_id', 'gender', 'zipcode', 'age_desc', 'occ_desc'])

# Reading movies file
movies = pd.read_csv('movies.csv', sep='\t', encoding='latin-1', usecols=['movie_id', 'title', 'genres'])

Let's take a look at the movies and ratings dataframes.

In [28]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

In [29]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

Also let's count the number of unique users and movies.

In [30]:
n_users = ratings.user_id.unique().shape[0]
n_movies = ratings.movie_id.unique().shape[0]
print ('Number of users = ' + str(n_users) + ' | Number of movies = ' + str(n_movies))

Number of users = 6040 | Number of movies = 3706


Now I want the format of my ratings matrix to be one row per user and one column per movie. To do so, I'll pivot *ratings* to get that and call the new variable *Ratings* (with a capital *R).

In [31]:
Ratings = ratings.pivot(index = 'user_id', columns ='movie_id', values = 'rating').fillna(0)
Ratings.head()

movie_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                               ...   
1          5.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
5          0.0   0.0   0.0   0.0   0.0   2.0   0.0   0.0   0.0   0.0  ...   

movie_id  3943  3944  3945  3946  3947  3948  3949  3950  3951  3952  
user_id                                                               
1          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5          0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 3706 columns]

Last but not least, I need to de-normalize the data (normalize by each users mean) and convert it from a dataframe to a numpy array.

In [32]:
R = Ratings.to_numpy()
user_ratings_mean = np.mean(R, axis = 1)
Ratings_demeaned = R - user_ratings_mean.reshape(-1, 1)

With my ratings matrix properly formatted and normalized, I'm ready to do some dimensionality reduction. But first, let's go over the math.

## Model-Based Collaborative Filtering
*Model-based Collaborative Filtering* is based on *matrix factorization (MF)* which has received greater exposure, mainly as an unsupervised learning method for latent variable decomposition and dimensionality reduction. Matrix factorization is widely used for recommender systems where it can deal better with scalability and sparsity than Memory-based CF:

* The goal of MF is to learn the latent preferences of users and the latent attributes of items from known ratings (learn features that describe the characteristics of ratings) to then predict the unknown ratings through the dot product of the latent features of users and items.
* When you have a very sparse matrix, with a lot of dimensions, by doing matrix factorization, you can restructure the user-item matrix into low-rank structure, and you can represent the matrix by the multiplication of two low-rank matrices, where the rows contain the latent vector.
* You fit this matrix to approximate your original matrix, as closely as possible, by multiplying the low-rank matrices together, which fills in the entries missing in the original matrix.

For example, let's check the sparsity of the ratings dataset:

In [33]:
sparsity = round(1.0 - len(ratings) / float(n_users * n_movies), 3)
print ('The sparsity level of MovieLens1M dataset is ' +  str(sparsity * 100) + '%')

The sparsity level of MovieLens1M dataset is 95.5%


## Support Vector Decomposition (SVD)
A well-known matrix factorization method is *Singular value decomposition (SVD)*. At a high level, SVD is an algorithm that decomposes a matrix $A$ into the best lower rank (i.e. smaller/simpler) approximation of the original matrix $A$. Mathematically, it decomposes A into a two unitary matrices and a diagonal matrix:

![svd](images/svd.png)

where $A$ is the input data matrix (users's ratings), $U$ is the left singular vectors (user "features" matrix), $\Sigma$ is the diagonal matrix of singular values (essentially weights/strengths of each concept), and  $V^{T}$ is the right singluar vectors (movie "features" matrix). $U$ and $V^{T}$ are column orthonomal, and represent different things. $U$ represents how much users "like" each feature and $V^{T}$ represents how relevant each feature is to each movie.

To get the lower rank approximation, I take these matrices and keep only the top $k$ features, which can be thought of as the underlying tastes and preferences vectors.

### Setting Up SVD
Scipy and Numpy both have functions to do the singular value decomposition. I'm going to use the Scipy function *svds* because it let's me choose how many latent factors I want to use to approximate the original ratings matrix (instead of having to truncate it after).

In [34]:
from scipy.sparse.linalg import svds
U, sigma, Vt = svds(Ratings_demeaned, k = 50)

As I'm going to leverage matrix multiplication to get predictions, I'll convert the $\Sigma$ (now are values) to the diagonal matrix form.

In [35]:
sigma = np.diag(sigma)

### Making Predictions from the Decomposed Matrices
I now have everything I need to make movie ratings predictions for every user. I can do it all at once by following the math and matrix multiply $U$, $\Sigma$, and $V^{T}$ back to get the rank $k=50$ approximation of $A$.

But first, I need to add the user means back to get the actual star ratings prediction.

In [36]:
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) + user_ratings_mean.reshape(-1, 1)

With the predictions matrix for every user, I can build a function to recommend movies for any user. I return the list of movies the user has already rated, for the sake of comparison.

In [37]:
preds = pd.DataFrame(all_user_predicted_ratings, columns = Ratings.columns)
preds.head()

movie_id      1         2         3         4         5         6     \
0         4.288861  0.143055 -0.195080 -0.018843  0.012232 -0.176604   
1         0.744716  0.169659  0.335418  0.000758  0.022475  1.353050   
2         1.818824  0.456136  0.090978 -0.043037 -0.025694 -0.158617   
3         0.408057 -0.072960  0.039642  0.089363  0.041950  0.237753   
4         1.574272  0.021239 -0.051300  0.246884 -0.032406  1.552281   

movie_id      7         8         9         10    ...      3943      3944  \
0        -0.074120  0.141358 -0.059553 -0.195950  ...  0.027807  0.001640   
1         0.051426  0.071258  0.161601  1.567246  ... -0.056502 -0.013733   
2        -0.131778  0.098977  0.030551  0.735470  ...  0.040481 -0.005301   
3        -0.049426  0.009467  0.045469 -0.111370  ...  0.008571 -0.005425   
4        -0.199630 -0.014920 -0.060498  0.450512  ...  0.110151  0.046010   

movie_id      3945      3946      3947      3948      3949      3950  \
0         0.026395 -0.022024 -0.085415  0.403529  0.105579  0.031912   
1        -0.010580  0.062576 -0.016248  0.155790 -0.418737 -0.101102   
2         0.012832  0.029349  0.020866  0.121532  0.076205  0.012345   
3        -0.008500 -0.003417 -0.083982  0.094512  0.057557 -0.026050   
4         0.006934 -0.015940 -0.050080 -0.052539  0.507189  0.033830   

movie_id      3951      3952  
0         0.050450  0.088910  
1        -0.054098 -0.140188  
2         0.015148 -0.109956  
3         0.014841 -0.034224  
4         0.125706  0.199244  

[5 rows x 3706 columns]

Now I write a function to return the movies with the highest predicted rating that the specified user hasn't already rated. Though I didn't use any explicit movie content features (such as genre or title), I'll merge in that information to get a more complete picture of the recommendations.

In [38]:
def recommend_movies(predictions, userID, movies, original_ratings, num_recommendations):

    # Get and sort the user's predictions
    user_row_number = userID - 1 # User ID starts at 1, not 0

    # Check if user_row_number is within the bounds of the predictions DataFrame
    if user_row_number >= predictions.shape[0]:
        print(f"User ID {userID} is not found in the data. Please provide a valid User ID.")
        return pd.DataFrame(), pd.DataFrame()  # Return empty DataFrames

    sorted_user_predictions = preds.iloc[user_row_number].sort_values(ascending=False) # User ID starts at 1

    # Get the user's data and merge in the movie information.
    user_data = original_ratings[original_ratings.user_id == (userID)]
    user_full = (user_data.merge(movies, how = 'left', left_on = 'movie_id', right_on = 'movie_id').
                     sort_values(['rating'], ascending=False)
                 )

    print ('User {0} has already rated {1} movies.'.format(userID, user_full.shape[0]))
    print ('Recommending highest {0} predicted ratings movies not already rated.'.format(num_recommendations))

    # Recommend the highest predicted rating movies that the user hasn't seen yet.
    recommendations = (movies[~movies['movie_id'].isin(user_full['movie_id'])].
         merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left',
               left_on = 'movie_id',
               right_on = 'movie_id').
         rename(columns = {user_row_number: 'Predictions'}).
         sort_values('Predictions', ascending = False).
                       iloc[:num_recommendations, :-1]
                      )

    return user_full, recommendations

Let's try to recommend 20 movies for user with ID 130.

In [39]:
already_rated, predictions = recommend_movies(preds, 130, movies, ratings, 20)

User 130 has already rated 135 movies.
Recommending highest 20 predicted ratings movies not already rated.


In [40]:
# Top 20 movies that User 130 has rated
already_rated.head(20)

user_id  movie_id  rating  timestamp  \
118      130      1220       5  977432036   
54       130      2490       5  977432233   
108      130      2879       5  977432036   
61       130      3527       5  977432056   
60       130       163       5  977432056   
59       130      2571       5  977431947   
88       130       368       5  977432262   
110      130      1210       5  977432036   
111      130      3753       5  977431354   
112      130      2951       5  977432108   
113      130      1214       5  977431970   
64       130      2662       5  977432084   
52       130      2476       5  977432193   
114      130      1215       5  977432036   
50       130      2393       5  977432283   
49       130      3406       5  977432108   
116      130      2019       5  977431921   
95       130       380       5  977432193   
46       130      2529       5  977432056   
45       130       110       5  977431970   

                                                 title  \
118                         Blues Brothers, The (1980)   
54                                      Payback (1999)   
108           Operation Condor (Feiying gaiwak) (1990)   
61                                     Predator (1987)   
60                                    Desperado (1995)   
59                                  Matrix, The (1999)   
88                                     Maverick (1994)   
110  Star Wars: Episode VI - Return of the Jedi (1983)   
111                                Patriot, The (2000)   
112                       Fistful of Dollars, A (1964)   
113                                       Alien (1979)   
64                       War of the Worlds, The (1953)   
52                             Heartbreak Ridge (1986)   
114                            Army of Darkness (1993)   
50                      Star Trek: Insurrection (1998)   
49                   Captain Horatio Hornblower (1951)   
116  Seven Samurai (The Magnificent Seven) (Shichin...   
95                                    True Lies (1994)   
46                           Planet of the Apes (1968)   
45                                   Braveheart (1995)   

                                    genres  
118                  Action|Comedy|Musical  
54                         Action|Thriller  
108                Action|Adventure|Comedy  
61                  Action|Sci-Fi|Thriller  
60                 Action|Romance|Thriller  
59                  Action|Sci-Fi|Thriller  
88                   Action|Comedy|Western  
110    Action|Adventure|Romance|Sci-Fi|War  
111                       Action|Drama|War  
112                         Action|Western  
113          Action|Horror|Sci-Fi|Thriller  
64                       Action|Sci-Fi|War  
52                              Action|War  
114  Action|Adventure|Comedy|Horror|Sci-Fi  
50                           Action|Sci-Fi  
49                    Action|Adventure|War  
116                           Action|Drama  
95         Action|Adventure|Comedy|Romance  
46                           Action|Sci-Fi  
45                        Action|Drama|War

In [41]:
# Top 20 movies that User 130 hopefully will enjoy
predictions

movie_id                                       title  \
1233      1304   Butch Cassidy and the Sundance Kid (1969)   
2826      2993                          Thunderball (1965)   
3242      3418                      Thelma & Louise (1991)   
2702      2858                      American Beauty (1999)   
1148      1197                  Princess Bride, The (1987)   
1865      2001                      Lethal Weapon 2 (1989)   
2136      2278                                Ronin (1998)   
699        733                            Rock, The (1996)   
1779      1912                         Out of Sight (1998)   
2935      3104                         Midnight Run (1988)   
1082      1127                           Abyss, The (1989)   
2223      2366                            King Kong (1933)   
2146      2288                           Thing, The (1982)   
1299      1375  Star Trek III: The Search for Spock (1984)   
1870      2006                   Mask of Zorro, The (1998)   
2608      2763             Thomas Crown Affair, The (1999)   
1506      1608                        Air Force One (1997)   
1474      1573                             Face/Off (1997)   
3452      3638                            Moonraker (1979)   
2824      2990                      Licence to Kill (1989)   

                                genres  
1233             Action|Comedy|Western  
2826                            Action  
3242                      Action|Drama  
2702                      Comedy|Drama  
1148   Action|Adventure|Comedy|Romance  
1865         Action|Comedy|Crime|Drama  
2136             Action|Crime|Thriller  
699          Action|Adventure|Thriller  
1779              Action|Crime|Romance  
2935     Action|Adventure|Comedy|Crime  
1082  Action|Adventure|Sci-Fi|Thriller  
2223           Action|Adventure|Horror  
2146     Action|Horror|Sci-Fi|Thriller  
1299           Action|Adventure|Sci-Fi  
1870          Action|Adventure|Romance  
2608                   Action|Thriller  
1506                   Action|Thriller  
1474            Action|Sci-Fi|Thriller  
3452             Action|Romance|Sci-Fi  
2824                            Action

These look like pretty good recommendations. It's good to see that, although I didn't actually use the genre of the movie as a feature, the truncated matrix factorization features "picked up" on the underlying tastes and preferences of the user. I've recommended some comedy, drama, and romance movies - all of which were genres of some of this user's top rated movies.

### Model Evaluation
Can't forget to evaluate our model, can we?

Instead of doing manually like the last time, I will use the *[Surprise](https://pypi.python.org/pypi/scikit-surprise)* library that provided various ready-to-use powerful prediction algorithms including (SVD) to evaluate its RMSE (Root Mean Squared Error) on the MovieLens dataset. It is a Python scikit building and analyzing recommender systems.

In [21]:
!pip install --upgrade scikit-surprise # Upgrade to latest version

In [44]:
# Import libraries from Surprise package
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate

# Load Reader library
reader = Reader()

# Load ratings dataset with Dataset library
data = Dataset.load_from_df(ratings[['user_id', 'movie_id', 'rating']], reader)

# Use the SVD algorithm.
svd = SVD()

# Compute the RMSE of the SVD algorithm using cross_validate.
cross_validate(svd, data, measures=['RMSE'], cv=5, verbose=True) # Use cross_validate and specify cv=5

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8752  0.8733  0.8733  0.8740  0.8712  0.8734  0.0013  
Fit time          24.14   16.50   16.99   16.55   17.55   18.35   2.92    
Test time         3.38    3.74    2.78    3.11    2.58    3.12    0.42    


{'test_rmse': array([0.87519246, 0.87334719, 0.87330783, 0.87400808, 0.8712458 ]),
 'fit_time': (24.1351580619812,
  16.503628492355347,
  16.99372625350952,
  16.545923233032227,
  17.550341844558716),
 'test_time': (3.382999897003174,
  3.741469621658325,
  2.7779529094696045,
  3.109924554824829,
  2.577648162841797)}

I get a mean *Root Mean Square Error* of 0.8734 which is pretty good. Let's now train on the dataset and arrive at predictions.

In [46]:
trainset = data.build_full_trainset()
svd.fit(trainset)

I'll pick again user with ID 130 and check the ratings he has given.

In [47]:
ratings[ratings['user_id'] == 130]

user_id  movie_id  rating  timestamp
16377      130      1249       4  977431294
16378      130      2989       5  977432262
16379      130       648       5  977431402
16380      130      3930       4  977431873
16381      130      3932       5  977431845
...        ...       ...     ...        ...
16507      130      2115       5  977432193
16508      130      1240       5  977431947
16509      130      3928       4  977431873
16510      130      1242       5  977431992
16511      130      2985       4  977432262

[135 rows x 4 columns]

Now let's use SVD to predict the rating that User with ID 130 will give to a random movie (let's say with Movie ID 1994).

In [48]:
svd.predict(130, 1994)

Prediction(uid=130, iid=1994, r_ui=None, est=3.7315107381868513, details={'was_impossible': False})

For movie with ID 1994, I get an estimated prediction of 3.731. The recommender system works purely on the basis of an assigned movie ID and tries to predict ratings based on how the other users have predicted the movie.

## Conclusion
In this notebook, I attempted to build a model-based Collaborative Filtering movie recommendation sytem based on latent features from a low rank matrix factorization method called SVD. As it captures the underlying features driving the raw data, it can scale significantly better to massive datasets as well as make better recommendations based on user's tastes.

However, we still likely lose some meaningful signals by using a low-rank approximation. Specifically, there's an interpretability problem as a singular vector specifies a linear combination of all input columns or rows. There's also a lack of sparsity when the singular vectors are quite dense. Thus, SVD approach is limited to linear projections.